File: Subjective-C/contexts_10/contexts_10/Subjective-C/OriginalSources/contexts_10.m
#import <Foundation/Foundation.h>
#include <stdlib.h>
#import "contexts_10.h" 
#import "ContextsInitializer.h" 

const static int ITER = 1000; 
const static int CALLS = 100; 

@implementation Foo
-(int) testS {
    NSLog(@"Error - base method called"); 
    return  -1;
}

@contexts Context1
-(int) testS {
    return 1;
}

@contexts Context2
-(int) testS {
    return 2;
}

@contexts Context3
-(int) testS {
    return 3;
}

@contexts Context4
-(int) testS {
    return 4;
}

@contexts Context5
-(int) testS {
    return 5;
}

@contexts Context6
-(int) testS {
    return 6;
}

@contexts Context7
-(int) testS {
    return 7;
}

@contexts Context8
-(int) testS {
    return 8;
}

@contexts Context9
-(int) testS {
    return 9;
}

@contexts Context10
-(int) testS {
    return 10;
}


-(double) testContexts {
    NSDate *start = [NSDate date];
    for(int i = 0; i < ITER; i++) {
        int num = 1 + arc4random_uniform(10);
        [[SCContextManager sharedContextManager] activateContextWithName:[NSString stringWithFormat:@"Context%d", num]];
        for(int j=1; j<=CALLS; j++) {
          [self testS];
        }
        [[SCContextManager sharedContextManager] deactivateContextWithName:[NSString stringWithFormat:@"Context%d", num]];
    }
    NSTimeInterval timeInterval = fabs([start timeIntervalSinceNow]);
    return timeInterval*1000;
} 

-(void) runs:(int) run {
    NSMutableString *data = [NSMutableString stringWithString:@"Iterations, Elapsed time in milliseconds \n "];
    for(int times = 1; times <= 100; times++) {
        double elapsed_time = [self testContexts];
        [data appendFormat:@"%d, %f \n", times, elapsed_time];
    }
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *fileName = [NSString stringWithFormat:@"%@/runs_contexts/10/test%d.csv",documentsDirectory, run];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (![fileManager fileExistsAtPath:fileName]) {
        BOOL success = [[NSString stringWithString: data] writeToFile:fileName atomically:YES encoding:NSUTF8StringEncoding error:nil];
        if (!success) {
            NSLog(@"Error writting file");
        }
    } else {
        NSFileHandle *fileHandle = [NSFileHandle fileHandleForWritingAtPath:fileName];
        NSData *textData = [data dataUsingEncoding:NSUTF8StringEncoding];
        [fileHandle writeData:textData];
        [fileHandle closeFile];
    }
}
@end

int main(int argc, const char * argv[]) {
    ContextsInitializer *ci = [[ContextsInitializer alloc] init];
    [ci initializeContexts];
    Foo *foo = [[Foo alloc] init];
    for(int run=1; run <=10; run++) {
        [foo runs:run];
     }
    return 0;
}

File: ContextTraits/contexts/contexts_10.js
require("underscore");
var Trait = require("./traits.js").Trait;
var cop = require("./context-traits.js");

var foo;
Foo = function() {};
Foo.prototype.test = function() {
    throw new Error('Base method needs to be specialized');
}
foo = new Foo();

Context1 = new cop.Context();
Adaptation1 = Trait ({
    test: function() {
        return 1;
    }
});
Context1.adapt(foo, Adaptation1); 

Context2 = new cop.Context();
Adaptation2 = Trait ({
    test: function() {
        return 2;
    }
});
Context2.adapt(foo, Adaptation2); 

Context3 = new cop.Context();
Adaptation3 = Trait ({
    test: function() {
        return 3;
    }
});
Context3.adapt(foo, Adaptation3); 

Context4 = new cop.Context();
Adaptation4 = Trait ({
    test: function() {
        return 4;
    }
});
Context4.adapt(foo, Adaptation4); 

Context5 = new cop.Context();
Adaptation5 = Trait ({
    test: function() {
        return 5;
    }
});
Context5.adapt(foo, Adaptation5); 

Context6 = new cop.Context();
Adaptation6 = Trait ({
    test: function() {
        return 6;
    }
});
Context6.adapt(foo, Adaptation6); 

Context7 = new cop.Context();
Adaptation7 = Trait ({
    test: function() {
        return 7;
    }
});
Context7.adapt(foo, Adaptation7); 

Context8 = new cop.Context();
Adaptation8 = Trait ({
    test: function() {
        return 8;
    }
});
Context8.adapt(foo, Adaptation8); 

Context9 = new cop.Context();
Adaptation9 = Trait ({
    test: function() {
        return 9;
    }
});
Context9.adapt(foo, Adaptation9); 

Context10 = new cop.Context();
Adaptation10 = Trait ({
    test: function() {
        return 10;
    }
});
Context10.adapt(foo, Adaptation10); 


function testContexts() {
    var i;
    var j;
    var ITER = 1000;
    var CALLS = 100;
    var beginning_time = new Date().getTime();
    for(i = 0; i < ITER; i++) {
        num = 1 + Math.floor(Math.random()*10);
        eval("Context"+num+".activate();");
        for(j=1; j<=CALLS; j++) {
          foo.test();
        }
        eval("Context"+num+".deactivate();");
    }
    var end_time = new Date().getTime();
    return end_time - beginning_time;
} 

function runs(run) {
    var fs = require("fs");
    var times;
    fs.writeFileSync("./runs_contexts/10/test"+run+".csv", "");
    fs.appendFileSync("./runs_contexts/10/test"+run+".csv", "iteration, Elapsed time in milliseconds \n");
    for(times = 1; times <= 100; times++) {
        var elapsed_time = testContexts();
        fs.appendFileSync("./runs_contexts/10/test"+run+".csv", times + "," + elapsed_time + "\n");
    }
}

var run;
for(run=1; run <=10; run++) {
    runs(run);
}

